File: expo-tucus/services/soundcloud.js
let cachedClientId = null;
let clientIdExpiry = 0;

const parseTrack = (t, artistId) => ({
  id: String(t.id),
  title: t.title || 'Unknown Track',
  artist: t.user?.username || 'Unknown Artist',
  artistId: artistId || String(t.user?.id || ''),
  artists: null,
  duration: (t.duration || 0) / 1000,
  thumbnail: t.artwork_url
    ? t.artwork_url.replace('-large.', '-t500x500.')
    : t.user?.avatar_url || '',
  source: 'soundcloud',
});

const FALLBACK_IDS = [
  'a3e059563d7fd3372b49b37f00a00bcf',
  '2t9loNQH90kzJcsF6jGLssKfJaNuISww',
  'b45b1aa10f1ac2941910a7f0d10f8e28',
  'MhFkNSjOPK5BCgSGJdQKyPhdOByGNrKj',
  'qHf7V3Y7rN0vJ9sD1hQ5yZ6kB0FJfR1e',
  'cYfVh7FmNH9sPjJqfUvL3kT2xWbR5nDg',
];

async function getClientId() {
  if (cachedClientId && Date.now() < clientIdExpiry) return cachedClientId;

  for (const id of FALLBACK_IDS) {
    try {
      const res = await fetch(
        `https://api-v2.soundcloud.com/search/tracks?q=test&limit=1&client_id=${id}`,
        { signal: AbortSignal.timeout(5000) }
      );
      if (res.ok) {
        cachedClientId = id;
        clientIdExpiry = Date.now() + 3600000;
        return id;
      }
    } catch {}
  }

  throw new Error('SoundCloud unavailable');
}

async function withClientId(fn) {
  try {
    const id = await getClientId();
    return await fn(id);
  } catch (e) {
    console.warn('SoundCloud error:', e.message);
    return null;
  }
}

export const soundcloud = {
  async search(query) {
    return (await withClientId(async (clientId) => {
      const enc = encodeURIComponent(query);
      const res = await fetch(
        `https://api-v2.soundcloud.com/search/tracks?q=${enc}&client_id=${clientId}&limit=20`,
        { signal: AbortSignal.timeout(8000) }
      );
      if (!res.ok) return [];
      const json = await res.json();
      return (json.collection || []).map(t => parseTrack(t));
    })) || [];
  },

  async getStreamUrl(trackId) {
    const clientId = await getClientId();
    const res = await fetch(
      `https://api-v2.soundcloud.com/tracks/${trackId}?client_id=${clientId}`,
      { signal: AbortSignal.timeout(8000) }
    );
    if (!res.ok) throw new Error('Track not found');
    const json = await res.json();
    const transcodings = json.media?.transcodings || [];
    const transcoding =
      transcodings.find(t => t.format?.protocol === 'progressive') ||
      transcodings[0];
    if (!transcoding?.url) throw new Error('No suitable format');

    const streamRes = await fetch(`${transcoding.url}?client_id=${clientId}`, {
      signal: AbortSignal.timeout(8000),
    });
    const streamJson = await streamRes.json();
    return streamJson.url || streamJson.location || '';
  },

  async getCharts(genre) {
    return (await withClientId(async (clientId) => {
      const enc = encodeURIComponent(genre);
      const res = await fetch(
        `https://api-v2.soundcloud.com/charts?kind=top&genre=soundcloud:genres:${enc}&client_id=${clientId}&limit=10`,
        { signal: AbortSignal.timeout(8000) }
      );
      if (!res.ok) return await this.search(`top ${genre}`);
      const json = await res.json();
      const tracks = (json.collection || [])
        .map(item => {
          const t = item.track || (item.id ? item : null);
          return t ? parseTrack(t) : null;
        })
        .filter(t => t && t.duration > 0 && t.duration <= 600);
      return tracks.length ? tracks : await this.search(`top ${genre}`);
    })) || [];
  },

  async getSimilar(trackId) {
    return (await withClientId(async (clientId) => {
      const res = await fetch(
        `https://api-v2.soundcloud.com/tracks/${trackId}/related?client_id=${clientId}&limit=20`,
        { signal: AbortSignal.timeout(8000) }
      );
      if (!res.ok) return [];
      const json = await res.json();
      return (json.collection || []).map(t => parseTrack(t));
    })) || [];
  },
};

File: src-tauri/src/soundcloud.rs
use crate::Track;
use serde_json::Value;

const TIMEOUT: u64 = 8;

static mut CACHED_CLIENT_ID: Option<String> = None;
static mut CLIENT_ID_EXPIRY: u64 = 0;

const FALLBACK_IDS: &[&str] = &[
    "a3e059563d7fd3372b49b37f00a00bcf",
    "2t9loNQH90kzJcsF6jGLssKfJaNuISww",
    "b45b1aa10f1ac2941910a7f0d10f8e28",
];

fn parse_track(t: &Value) -> Option<Track> {
    let id = t.get("id")?.as_i64()?.to_string();
    let title = t.get("title")?.as_str()?.to_string();
    let artist = t
        .pointer("/user/username")?
        .as_str()?
        .to_string();
    let artwork = t.get("artwork_url").and_then(|v| v.as_str());
    let user_avatar = t.pointer("/user/avatar_url").and_then(|v| v.as_str());
    let thumbnail = artwork
        .map(|a| a.replace("-large.", "-t500x500."))
        .or_else(|| user_avatar.map(String::from));
    let duration_ms = t.get("duration")?.as_f64()?;
    Some(Track {
        id,
        title,
        artist,
        source: "soundcloud".into(),
        thumbnail,
        duration: Some(duration_ms / 1000.0),
    })
}

async fn get_client_id() -> Result<String, String> {
    let now = std::time::SystemTime::now()
        .duration_since(std::time::UNIX_EPOCH)
        .unwrap()
        .as_millis() as u64;

    unsafe {
        if let Some(ref id) = CACHED_CLIENT_ID {
            if now < CLIENT_ID_EXPIRY {
                return Ok(id.clone());
            }
        }
    }

    let client = reqwest::Client::new();
    for id in FALLBACK_IDS {
        let url = format!(
            "https://api-v2.soundcloud.com/search/tracks?q=test&limit=1&client_id={}",
            id
        );
        if let Ok(resp) = client
            .get(&url)
            .timeout(std::time::Duration::from_secs(5))
            .send()
            .await
        {
            if resp.status().is_success() {
                let id_owned = id.to_string();
                unsafe {
                    CACHED_CLIENT_ID = Some(id_owned.clone());
                    CLIENT_ID_EXPIRY = now + 3600000;
                }
                return Ok(id_owned);
            }
        }
    }
    Err("SoundCloud unavailable".into())
}

pub async fn search(query: &str) -> Result<Vec<Track>, String> {
    let client_id = get_client_id().await?;
    let client = reqwest::Client::new();
    let url = format!(
        "https://api-v2.soundcloud.com/search/tracks?q={}&client_id={}&limit=20",
        urlencoding::encode(query),
        client_id
    );
    let resp = client
        .get(&url)
        .timeout(std::time::Duration::from_secs(TIMEOUT))
        .send()
        .await
        .map_err(|e| e.to_string())?;
    let json: Value = resp.json().await.map_err(|e| e.to_string())?;
    let tracks = json["collection"]
        .as_array()
        .map(|arr| arr.iter().filter_map(parse_track).collect())
        .unwrap_or_default();
    Ok(tracks)
}

pub async fn get_stream_url(track_id: &str) -> Result<String, String> {
    let client_id = get_client_id().await?;
    let client = reqwest::Client::new();
    let url = format!(
        "https://api-v2.soundcloud.com/tracks/{}?client_id={}",
        track_id, client_id
    );
    let resp = client
        .get(&url)
        .timeout(std::time::Duration::from_secs(TIMEOUT))
        .send()
        .await
        .map_err(|e| e.to_string())?;
    let json: Value = resp.json().await.map_err(|e| e.to_string())?;
    let transcodings = json["media"]["transcodings"]
        .as_array()
        .ok_or("No transcodings")?;
    let transcoding = transcodings
        .iter()
        .find(|t| t["format"]["protocol"].as_str() == Some("progressive"))
        .or_else(|| transcodings.first())
        .ok_or("No suitable format")?;
    let t_url = transcoding["url"].as_str().ok_or("No URL")?;

    let stream_resp = client
        .get(format!("{}?client_id={}", t_url, client_id))
        .timeout(std::time::Duration::from_secs(TIMEOUT))
        .send()
        .await
        .map_err(|e| e.to_string())?;
    let stream_json: Value = stream_resp.json().await.map_err(|e| e.to_string())?;
    stream_json["url"]
        .as_str()
        .or_else(|| stream_json["location"].as_str())
        .map(String::from)
        .ok_or("No stream URL".into())
}
